elytron/WFLY-17143-request-uri-parameters.adoc

```diff
@@ -0,0 +1,196 @@
+
+** The JWT must be signed and then encrypted to make a nested JWT and the key used to encrypt the JWT must be the realm public key shared by the OpenID provided. The key algorithm must be the same as the encrypting algorithm. Signing and encrypting algorithms can be of different types (i. e. ES-256 for signing and RSA-OAEP for encrypting is allowed). The realm key for encrypting the JWT can be obtained by sending a GET request to the `jwksUrl` obtained from the `discoveryUrl`. 
+
+* The authentication request can send the Request Object in one of two ways: 
+
+** By value: The `request` parameter is added to the query and its value is the Request Object itself. 
+
+** The request string can be added to the auth request by adding the whole string: 
+
+```
+https://idsvr.example.com/oauth/v2/oauth-authorize?
+&client_id=client-one
+&response_type=code
+&scope=openid
+&request=y9Lqv4fCp6Ei-u2-ZCKq83YvbFEk6JMs...
+```
+** By reference: This is useful when the `request` string is too large and a number of other https://openid.net/specs/openid-connect-core-1_0.html#RequestUriRationale[reasons]. In this case, the Request Object is passed as a reference parameter string using the `request_uri` parameter. The url references the Request Object, which the OpenID provider can use to download the Request Object itself. The entire Request_URI MUST NOT exceed 512 ASCII characters. 
+
+** The `request_uri` value can be obtained by sending POST request to the `pushed_authentication_request_endpoint` of the OpenID provider. This can be also obtained using the discoveryUrl. The content for the POST request contains the value of the signed and/or encrypted Request Object and the content type for this request is "application/x-www-form-urlencoded". 
+
+** A successful request will return 201 http code along with a JWT, which contains the value of the `request_uri` and its expiry time.  
+
+** When included in the auth request, the request_uri would be specified as: 
+    
+```
+    https://server.example.com/authorize?
+    response_type=code%20id_token
+    &client_id=s6BhdRkqt3
+    &request_uri=https%3A%2F%2Fclient.example.org%2Frequest.jwt
+    %23GkurKxf5T0Y-mnPFCHqWOMiZi4VS138cQO_V7PZHAdM
+    &state=af0ifjsldkj&nonce=n-0S6_WzA2Mj
+    &scope=openid
+```
+
+** The base64-encoded SHA-256 hash `23GkurKxf5T0Y-mnPFCHqWOMiZi4VS138cQO_V7PZHAdM` will allow the provider to cache the token for future use. It is a hash of the contents of the Request Object.
+
+* According to the OpenID docs, if the same parameter exists both in the Request Object and the OAuth Authorization Request parameters, the parameter in the Request Object is used.
+
+* The feature must deal with cases where the OpenId Provider does not support request parameters by recognizing `request_not_supported` error and dealing with it accordingly.
```

Comment:
@PrarthonaPaul `by recognizing `request_not_supported` error and dealing with it accordingly.` what will be the approach for this situation? Will some error be thrown that says that request parameters are not supported and only oauth2 format is supported?

Reply:
For now I have coded it so that if either request or request_uri is not supported, then we send it as usual.
But this may not be as secure. So, in that case, if it makes more sense to, I can change it to throw a runtime error saying that request/request_uri is not supported by the OpenID provider
Which one do you think would be more appropriate?

Reply:
@PrarthonaPaul I think we should log this and then document that the request and the request_uri methods are used only if available with the OpenID provider, if not available then the warning message will be logged.